[FEATURE] More complex RBP example including variables

Shinnnyshinshin opened the pull request:
Changes proposed in this pull request:
- Adding a more complex example to the Notebook using multiple batches, multiple ParameterBuilders, ExpectationConfigurationBuilders and introducing the use of variables.

The goal of the example is to build a RuleBasedProfiler that outputs an ExpectationSuite containing 2 Expectation types
- `expect_column_min_to_be_between`
- `expect_column_max_to_be_between`

for 2 columns in our taxi_data dataset
- `fare_amount`
- `tip_amount`

with the `min_value` and `max_value` parameters for each of the Expectations estimated over 12 Batches of `taxi_data`, for a total of 4 Expectations.

To estimate the parameters, we will be using a NumericMetricRangeMultiBatchParameterBuilder, which is able to provide range estimations across Batches using sampling methods. We will also be using a variables dictionary to share defined variables across Rule components like DomainBuilders, ParameterBuilders and ExpectationConfigurationBuilders.

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

Commit:
first push

Commit:
Update BasicExample_RBP_Instantiation_and_running.ipynb

Commit:
before clean up

Commit:
update before cleaning up

Commit:
submitting the notebook

Commit:
cleaning up yaml

Commit:
Merge branch 'develop' into feature/GREAT-500/add-notebooks-for-rbp

Commit:
Merge branch 'develop' into feature/GREAT-500/add-notebooks-for-rbp

Commit:
first update

Commit:
Merge branch 'feature/GREAT-500/add-notebooks-for-rbp' of https://github.com/great-expectations/great_expectations into feature/GREAT-500/add-notebooks-for-rbp

* 'feature/GREAT-500/add-notebooks-for-rbp' of https://github.com/great-expectations/great_expectations:
  [FEATURE] Use more granular requirements-dev-xxx.txt files (#4327)
  standardise on include and exclude columns list for profiler (#4424)
  [MAINTENANCE] Continue chipping away at warnings (#4422)
  [MAINTENANCE] Add docstrings and type hints to `Anonymizer` (#4419)
  Replace "sampling_method" with "estimator", since the latter is a more fitting name for the parameter. (#4420)
  [BUGFIX] Adding `--spark` flag back to `azure-pipelines.yml` compatibility_matrix stage.  (#4418)
  [MAINTENANCE] Refactor out unnecessary Anonymizer child classes (#4408)
  [BUGFIX] Pin `pytest-azurepipelines` due to breaking changes made in latest release

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Rachel-Reverie <94694058+Rachel-Reverie@users.noreply.github.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Rachel-Reverie <94694058+Rachel-Reverie@users.noreply.github.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Rachel-Reverie <94694058+Rachel-Reverie@users.noreply.github.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Rachel-Reverie <94694058+Rachel-Reverie@users.noreply.github.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Rachel-Reverie <94694058+Rachel-Reverie@users.noreply.github.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Rachel-Reverie <94694058+Rachel-Reverie@users.noreply.github.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Rachel-Reverie <94694058+Rachel-Reverie@users.noreply.github.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Rachel-Reverie <94694058+Rachel-Reverie@users.noreply.github.com>

Commit:
final push

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Rachel-Reverie <94694058+Rachel-Reverie@users.noreply.github.com>

Commit:
Merge branch 'develop' into feature/GREAT-500/add-notebooks-for-rbp

Commit:
Update BasicExample_RBP_Instantiation_and_running.ipynb

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Anthony Burdi <anthony@superconductive.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Anthony Burdi <anthony@superconductive.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Anthony Burdi <anthony@superconductive.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Anthony Burdi <anthony@superconductive.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Anthony Burdi <anthony@superconductive.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Anthony Burdi <anthony@superconductive.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Anthony Burdi <anthony@superconductive.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Anthony Burdi <anthony@superconductive.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Anthony Burdi <anthony@superconductive.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Anthony Burdi <anthony@superconductive.com>

Commit:
Merge branch 'develop' into feature/GREAT-500/add-notebooks-for-rbp

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Anthony Burdi <anthony@superconductive.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Anthony Burdi <anthony@superconductive.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Anthony Burdi <anthony@superconductive.com>

Commit:
Update tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb

Co-authored-by: Anthony Burdi <anthony@superconductive.com>

Commit:
push before last

Commit:
Revert "push before last"

This reverts commit c79308208a029c51b959e43ecf707dfcfcbebc3e.

Commit:
Revert "Revert "push before last""

This reverts commit fbfa23db4c6327810ce7cda8aae5138c70573797.

Commit:
push

Commit:
final

Commit:
final touch

Commit:
removed bugfix note

Commit:
Adding complex example

alexsherstinsky commented on tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb:
```diff
@@ -758,6 +758,361 @@
+   "source": [
+    "variables: dict = {\n",
+    "    \"multi_batch_batch_request\": multi_batch_batch_request,\n",
+    "    \"estimator_name\": \"oneshot\",\n",
```
@Shinnnyshinshin This "oneshot" is not a real estimator -- it is provided for testing purposes only.  Would using `bootstrap` work for your purposes at all?  If not, then I guess "oneshot" is OK -- as long as we put a disclaimer.  Thanks.

Shinnnyshinshin replied:
It would, but we could run into some trouble with the tests sometimes resulting in a warning because of the non-deterministic nature of the `bootstrap`.  I think we should be ok here though

alexsherstinsky commented on tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb:
```diff
@@ -758,6 +758,361 @@
+   "source": [
+    "### Instantiating `RuleBasedProfiler` with `variables`\n",
+    "\n",
+    "Pass the `variables` directionary into the `RuleBasedProfiler` constructor. "
```
@Shinnnyshinshin "directionary"?  Or "dictionary"?..

Shinnnyshinshin replied:
😓

alexsherstinsky commented on tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb:
```diff
@@ -758,6 +758,361 @@
+    "    metric_domain_kwargs=\"$domain.domain_kwargs\",\n",
+    "    false_positive_rate=\"$variables.false_positive_rate\",\n",
+    "    data_context=data_context,\n",
+    "    batch_request=\"$variables.multi_batch_batch_request\",\n",
```
@Shinnnyshinshin Any chance to keep the order of arguments the way they are in the class definition (for elegance/readability purposes, that is) -- thanks!

alexsherstinsky reviewed (changes requested):
@Shinnnyshinshin Good stuff -- a couple of minor questions/requests.  In general, it would be easier to review the notebook in its proper executable rendering (hard to read its source code), hence, I would be pleased to pair on it.  Thanks!

Rachel-Reverie reviewed (approved):
Changes to the markdown sections look good to me; the Notebook executed all cells without errors, as well.

Commit:
adding integration test for notebook

Commit:
Merge branch 'develop' into feature/GREAT-676/adding-variables-examples

* develop:
  [BUGFIX] Ensure that atomic prescriptive renderers have proper `mostly_pct` param in payload (#4438)
  [MAINTENANCE] Update broken snippet threshold in embedded Docusaurus reference validation script
  [FEATURE] Notebooks for RBP Examples (#4396)
  updated release candidate for 0.14.11 (#4454)
  [BUGFIX] Fix packaging test
  [BUGFIX] Removes potentially sensitive webhook URLs from logging (#4440)
  release candidate for 0.14.11 (#4448)
  Add any dependencies we import directly, but don't have as explicit requirements (#4447)
  [FEATURE] Rule-Based Profiler: Implement Utilities for getting all available parameter node names and objects resident in memory (#4442)

Conflicts:
	tests/test_fixtures/rule_based_profiler/example_notebooks/BasicExample_RBP_Instantiation_and_running.ipynb
	tests/test_fixtures/rule_based_profiler/example_notebooks/great_expectations/great_expectations.yml

Commit:
Update BasicExample_RBP_Instantiation_and_running.ipynb

Commit:
Merge branch 'develop' into feature/GREAT-676/adding-variables-examples

Shinnnyshinshin commented:
Thank you @alexsherstinsky for the review. I've added the changes you mentioned, and adjusted the False-positive rate to be 5% which seemed to help with the warning we were getting sometimes. 

Could I get a re-review?

Regarding being able to read the Notebook in its rendered form, if you go to `View File` in the web interface, you should be able view the rendered version.

Commit:
Merge branch 'develop' into feature/GREAT-676/adding-variables-examples

Commit:
Merge branch 'develop' into feature/GREAT-676/adding-variables-examples

Commit:
updated after code changes

Commit:
Update BasicExample_RBP_Instantiation_and_running.ipynb

Commit:
Merge branch 'develop' into feature/GREAT-676/adding-variables-examples

alexsherstinsky reviewed (approved):
LGTM -- thank you for this, @Shinnnyshinshin !

Commit:
Merge branch 'develop' into feature/GREAT-676/adding-variables-examples
